src/StaticWebAssetsSdk/Tasks/ApplyCompressionNegotiation.cs

```diff
@@ -184,26 +125,187 @@ public override bool Execute()
+    {
+        var result = new Dictionary<string, List<StaticWebAssetEndpoint>>(endpointsByAsset.Count);
+
+        foreach (var endpointsList in endpointsByAsset.Values)
+        {
+            foreach (var endpoint in endpointsList)
+            {
+                if (!result.TryGetValue(endpoint.Route, out var routeEndpoints))
+                {
+                    routeEndpoints = new List<StaticWebAssetEndpoint>(5);
+                    result[endpoint.Route] = routeEndpoints;
+                }
+                routeEndpoints.Add(endpoint);
+            }
+        }
+
+        return result;
+    }
+
+    private static StaticWebAssetEndpointResponseHeader[] GetOrCreateCompressionHeaders(Dictionary<string, StaticWebAssetEndpointResponseHeader[]> compressionHeadersByEncoding, StaticWebAsset compressedAsset)
+    {
+        if (!compressionHeadersByEncoding.TryGetValue(compressedAsset.AssetTraitValue, out var compressionHeaders))
+        {
+            compressionHeaders = CreateCompressionHeaders(compressedAsset);
+            compressionHeadersByEncoding.Add(compressedAsset.AssetTraitValue, compressionHeaders);
+        }
+
+        return compressionHeaders;
+    }
+
+    private static StaticWebAssetEndpointResponseHeader[] CreateCompressionHeaders(StaticWebAsset compressedAsset) =>
+        [
+            new()
+            {
+                Name = "Content-Encoding",
+                Value = compressedAsset.AssetTraitValue
+            },
+            new()
+            {
+                Name = "Vary",
+                Value = "Content-Encoding"
+            }
+        ];
+
+    private StaticWebAssetEndpoint CreateUpdatedEndpoint(
+        StaticWebAsset compressedAsset,
+        string quality,
+        StaticWebAssetEndpoint compressedEndpoint,
+        HashSet<string> compressedHeaders,
+        StaticWebAssetEndpoint relatedEndpointCandidate)
+    {
+        Log.LogMessage(MessageImportance.Low, "Processing related endpoint '{0}'", relatedEndpointCandidate.Route);
+        var encodingSelector = new StaticWebAssetEndpointSelector
+        {
+            Name = "Content-Encoding",
+            Value = compressedAsset.AssetTraitValue,
+            Quality = quality
+        };
+        Log.LogMessage(MessageImportance.Low, "  Created Content-Encoding selector for compressed asset '{0}' with size '{1}' is '{2}'", encodingSelector.Value, encodingSelector.Quality, relatedEndpointCandidate.Route);
+        var endpointCopy = new StaticWebAssetEndpoint
+        {
+            AssetFile = compressedAsset.Identity,
+            Route = relatedEndpointCandidate.Route,
+            Selectors = [
+                ..relatedEndpointCandidate.Selectors,
+                            encodingSelector
+            ],
+            EndpointProperties = relatedEndpointCandidate.EndpointProperties
+        };
+        var headers = new List<StaticWebAssetEndpointResponseHeader>(7);
```

Comment:
Why 7? How would we know when to update this in the future?

Reply:
It's the number of headers we add at most by default.